File: app/src/main/java/tabletop/velocic/com/worldforgerpgtools/GeneratorCreationFragment.java
package tabletop.velocic.com.worldforgerpgtools;

import android.app.Activity;
import android.content.Intent;
import android.os.Bundle;
import android.support.v4.app.Fragment;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Button;
import android.widget.EditText;
import android.widget.TextView;

import tabletop.velocic.com.worldforgerpgtools.GeneratorDeserializer.GeneratorImporter;

public class GeneratorCreationFragment extends android.support.v4.app.Fragment
{
    public static final String BACK_STACK_GENERATOR_CREATION_FRAGMENT = "tabletop.velocic.com.worldforgerpgtools.GeneratorCreationFragment";

    private static final int REQUEST_NEW_CATEGORY_PATH = 0;

    private EditText newGeneratorName;
    private TextView newGeneratorCategoryName;
    private Button createNewResultEntryButton;
    private Button submitGeneratorButton;

    public static GeneratorCreationFragment newInstance()
    {
        GeneratorCreationFragment fragment = new GeneratorCreationFragment();

        return fragment;
    }

    @Override
    public void onCreate(Bundle savedInstanceState)
    {
        super.onCreate(savedInstanceState);
    }

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState)
    {
        View view = inflater.inflate(R.layout.fragment_create_generator, container, false);

        newGeneratorName = (EditText) view.findViewById(R.id.edit_text_create_generator_name);
        newGeneratorCategoryName = (TextView) view.findViewById(R.id.edit_text_create_generator_category);
        createNewResultEntryButton = (Button) view.findViewById(R.id.button_add_generator_possible_result);
        submitGeneratorButton = (Button) view.findViewById(R.id.button_submit_new_generator);

        newGeneratorCategoryName.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v)
            {
                onNewGeneratorCategoryNameClicked(v);
            }
        });

        return view;
    }

    @Override
    public void onActivityResult(int requestCode, int resultCode, Intent data)
    {
        super.onActivityResult(requestCode, resultCode, data);

        if (resultCode != Activity.RESULT_OK) {
            return;
        }

        if (requestCode == REQUEST_NEW_CATEGORY_PATH) {
            //Can't set the TextView directly, as this function is called before
            //onResume, and the TextView state gets lost.
            setArguments(data.getExtras());
        }
    }

    @Override
    public void onResume()
    {
        super.onResume();

        Bundle fragmentArgs = getArguments();

        if (fragmentArgs != null) {
            newGeneratorCategoryName.setText(fragmentArgs.getString(GeneratorCategorySelectionFragment.EXTRA_SELECTED_CATEGORY));
        }
    }

    private void onNewGeneratorCategoryNameClicked(View v)
    {
        TextView view = (TextView) v;
        Fragment generatorCategorySelectionFragment = GeneratorCategorySelectionFragment.newInstance(view.getText().toString());

        generatorCategorySelectionFragment.setTargetFragment(GeneratorCreationFragment.this, REQUEST_NEW_CATEGORY_PATH);

        getActivity().getSupportFragmentManager().beginTransaction()
            .replace(R.id.fragment_container, generatorCategorySelectionFragment)
            .addToBackStack(BACK_STACK_GENERATOR_CREATION_FRAGMENT)
            .commit();
    }
}

File: app/src/main/java/tabletop/velocic/com/worldforgerpgtools/GeneratorCreationFragment.kt
package tabletop.velocic.com.worldforgerpgtools

import android.app.Activity
import android.content.Intent
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.TextView
import kotlinx.android.synthetic.main.activity_fragment.view.*
import kotlinx.android.synthetic.main.fragment_create_generator.view.*

class GeneratorCreationFragment : android.support.v4.app.Fragment()
{
    private var newGeneratorCategoryName: TextView = TextView(context)


    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?) : View
    {
        val view = inflater.inflate(R.layout.fragment_create_generator, container, false)

        newGeneratorCategoryName = view.edit_text_create_generator_category

        //TODO: Finish behavior for this screen after kotlin rewrite is complete
        val newGeneratorName = view.edit_text_create_generator_name
        val createNewResultEntryButton = view.button_add_generator_possible_result
        val submitGeneratorButton = view.button_submit_new_generator

        newGeneratorCategoryName.setOnClickListener(::onNewGeneratorCategoryNameClicked)

        return view
    }

    override fun onResume()
    {
        super.onResume()

        arguments?.let {
            newGeneratorCategoryName.text = it.getString(GeneratorCategorySelectionFragment.EXTRA_SELECTED_CATEGORY)
        }
    }

    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        super.onActivityResult(requestCode, resultCode, data)

        if (resultCode != Activity.RESULT_OK) {
            return;
        }

        if (requestCode == REQUEST_NEW_CATEGORY_PATH) {
            //Can't set the TextView directly, as this function is called before
            //onResume, and the TextView state gets lost
            arguments = data?.extras
        }
    }

    private fun onNewGeneratorCategoryNameClicked(v: View)
    {
        val textView = v as TextView
        val generatorCategorySelectionFragment = GeneratorCategorySelectionFragment.newInstance(textView.text.toString())

        activity?.supportFragmentManager?.beginTransaction()?.run {
            replace(R.id.fragment_container, generatorCategorySelectionFragment)
            addToBackStack(BACK_STACK_GENERATOR_CREATION_FRAGMENT)
            commit()
        }
    }

    companion object {
        const val BACK_STACK_GENERATOR_CREATION_FRAGMENT = "tabletop.velocic.com.worldforgerpgtools.GeneratorCreationFragment"
        private const val REQUEST_NEW_CATEGORY_PATH = 0

        fun newInstance() : GeneratorCreationFragment
        {
            return GeneratorCreationFragment()
        }
    }
}